File: app/src/main/java/com/example/service/TradingStrategyService.kt
package com.example.service

import com.example.model.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sqrt

class TradingStrategyService(
  private val client: OkHttpClient = OkHttpClient.Builder()
    .connectTimeout(8, TimeUnit.SECONDS)
    .readTimeout(8, TimeUnit.SECONDS)
    .build()
) {
  companion object {
    const val TESTNET_BASE_URL = "https://testnet.binance.vision"
  }

  suspend fun fetchKlines(symbol: String, interval: String = "5m", limit: Int = 60): List<List<Double>> = withContext(Dispatchers.IO) {
    val url = "$TESTNET_BASE_URL/api/v3/klines?symbol=${symbol.uppercase().trim()}&interval=$interval&limit=$limit"
    val req = Request.Builder().url(url).get().build()
    try {
      client.newCall(req).execute().use { res ->
        val body = res.body?.string().orEmpty()
        if (res.isSuccessful) {
          val array = JSONArray(body)
          val list = mutableListOf<List<Double>>()
          for (i in 0 until array.length()) {
            val candle = array.getJSONArray(i)
            list.add(
              listOf(
                candle.getDouble(0), // openTime
                candle.getString(1).toDoubleOrNull() ?: 0.0, // open
                candle.getString(2).toDoubleOrNull() ?: 0.0, // high
                candle.getString(3).toDoubleOrNull() ?: 0.0, // low
                candle.getString(4).toDoubleOrNull() ?: 0.0, // close
                candle.getString(5).toDoubleOrNull() ?: 0.0, // volume
              )
            )
          }
          list
        } else {
          emptyList()
        }
      }
    } catch (_: Exception) {
      emptyList()
    }
  }

  fun evaluate(
    symbol: String,
    currentPrice: Double,
    klines: List<List<Double>>,
    config: StrategyRiskConfig
  ): SignalData {
    val closes = klines.map { it[4] }
    val volumes = klines.map { it[5] }

    val rsi = calculateRsi(closes, 14)
    val emaFast = calculateEma(closes, 9)
    val emaSlow = calculateEma(closes, 21)

    // Честный расчет MACD (EMA12/EMA26, сигнальная линия EMA9 от ряда MACD)
    val emaFastSeries = calculateEmaSeries(closes, 12)
    val emaSlowSeries = calculateEmaSeries(closes, 26)
    val macdSeries = emaFastSeries.zip(emaSlowSeries) { f, s -> f - s }
    val signalSeries = calculateEmaSeries(macdSeries, 9)
    val macdLine = macdSeries.lastOrNull() ?: 0.0
    val macdSignal = signalSeries.lastOrNull() ?: 0.0
    val macdHist = macdLine - macdSignal

    val atr = calculateAtr(klines, config.atrPeriod)
    val avgVol = if (volumes.isNotEmpty()) volumes.takeLast(20).average() else 100.0
    val curVol = volumes.lastOrNull() ?: 120.0

    val bbWindow = if (closes.size >= 20) closes.takeLast(20) else closes
    val bbMid = if (bbWindow.isNotEmpty()) bbWindow.average() else currentPrice
    val std = if (bbWindow.isNotEmpty()) sqrt(bbWindow.map { (it - bbMid).pow(2) }.average()) else 10.0
    val bbUpper = bbMid + (std * 2)
    val bbLower = bbMid - (std * 2)

    val ind = IndicatorValues(
      rsi = rsi,
      emaFast = emaFast,
      emaSlow = emaSlow,
      macdLine = macdLine,
      macdSignal = macdSignal,
      macdHist = macdHist,
      currentVolume = curVol,
      avgVolume = avgVol,
      bbUpper = bbUpper,
      bbMiddle = bbMid,
      bbLower = bbLower,
      atr = atr
    )

    var longScore = 0
    var shortScore = 0
    val matched = mutableListOf<String>()

    if (emaFast > emaSlow) {
      longScore += 25
      matched.add("EMA9 > EMA21 (Бычий импульс)")
    } else {
      shortScore += 25
      matched.add("EMA9 < EMA21 (Медвежий тренд)")
    }

    if (rsi <= 35.0) {
      longScore += 25
      matched.add("RSI (%.1f) в зоне перепроданности".format(rsi))
    } else if (rsi >= 65.0) {
      shortScore += 25
      matched.add("RSI (%.1f) в зоне перекупленности".format(rsi))
    } else {
      longScore += 10
      shortScore += 10
    }

    // Сигнал по гистограмме MACD (пересечение MACD и сигнальной линии)
    if (macdHist >= 0) {
      longScore += 20
      matched.add("MACD-гистограмма в положительной зоне")
    } else {
      shortScore += 20
      matched.add("MACD-гистограмма в отрицательной зоне")
    }

    if (curVol > avgVol) {
      longScore += 15
      shortScore += 15
      matched.add("Объём выше среднего за 20 периодов")
    }

    if (currentPrice <= bbLower * 1.01) {
      longScore += 15
      matched.add("Цена у нижней границы Боллинджера (L: %.1f)".format(bbLower))
    } else if (currentPrice >= bbUpper * 0.99) {
      shortScore += 15
      matched.add("Цена у верхней границы Боллинджера (U: %.1f)".format(bbUpper))
    }

    val action = when {
      longScore >= config.minScoreThreshold && longScore > shortScore -> SignalAction.BUY_LONG
      shortScore >= config.minScoreThreshold && shortScore > longScore -> SignalAction.SELL_SPOT
      else -> SignalAction.HOLD
    }

    val finalScore = max(longScore, shortScore).coerceIn(0, 100)

    // Расчет Stop-Loss с поддержкой волатильного ATR
    val sl = if (config.useAtrSlTp && atr > 0.0) {
      if (action == SignalAction.BUY_LONG) {
        currentPrice - (atr * config.atrSlMultiplier)
      } else {
        currentPrice + (atr * config.atrSlMultiplier)
      }
    } else {
      if (action == SignalAction.BUY_LONG) {
        currentPrice * (1.0 - (config.stopLossPercent / 100.0))
      } else {
        currentPrice * (1.0 + (config.stopLossPercent / 100.0))
      }
    }

    // Расчет Take-Profit с поддержкой волатильного ATR
    val tp = if (config.useAtrSlTp && atr > 0.0) {
      if (action == SignalAction.BUY_LONG) {
        currentPrice + (atr * config.atrTpMultiplier)
      } else {
        currentPrice - (atr * config.atrTpMultiplier)
      }
    } else {
      if (action == SignalAction.BUY_LONG) {
        currentPrice * (1.0 + (config.takeProfitPercent / 100.0))
      } else {
        currentPrice * (1.0 - (config.takeProfitPercent / 100.0))
      }
    }

    return SignalData(
      symbol = symbol,
      currentPrice = currentPrice,
      indicators = ind,
      action = action,
      score = finalScore,
      matchedConditions = matched,
      recommendedStopLoss = sl,
      recommendedTakeProfit = tp
    )
  }

  private fun calculateEmaSeries(prices: List<Double>, period: Int): List<Double> {
    if (prices.isEmpty()) return emptyList()
    if (prices.size < period) {
      var runningSum = 0.0
      return prices.mapIndexed { idx, p ->
        runningSum += p
        runningSum / (idx + 1)
      }
    }
    val result = ArrayList<Double>(prices.size)
    val k = 2.0 / (period + 1.0)
    var initialSum = 0.0
    for (i in 0 until period) {
      initialSum += prices[i]
      result.add(initialSum / (i + 1))
    }
    var currentEma = initialSum / period
    for (i in period until prices.size) {
      currentEma = (prices[i] * k) + (currentEma * (1.0 - k))
      result.add(currentEma)
    }
    return result
  }

  fun calculateAtr(klines: List<List<Double>>, period: Int = 14): Double {
    if (klines.isEmpty()) return 0.0
    if (klines.size == 1) {
      val high = klines[0][2]
      val low = klines[0][3]
      return max(high - low, 0.0)
    }

    val trList = ArrayList<Double>(klines.size)
    for (i in klines.indices) {
      val high = klines[i][2]
      val low = klines[i][3]
      if (i == 0) {
        trList.add(max(high - low, 0.0))
      } else {
        val prevClose = klines[i - 1][4]
        val tr = max(high - low, max(kotlin.math.abs(high - prevClose), kotlin.math.abs(low - prevClose)))
        trList.add(tr)
      }
    }

    if (trList.size <= period) {
      return trList.average()
    }

    var atr = trList.take(period).average()
    for (i in period until trList.size) {
      atr = (atr * (period - 1) + trList[i]) / period
    }
    return atr
  }

  private fun calculateEma(prices: List<Double>, period: Int): Double {
    if (prices.isEmpty()) return 0.0
    if (prices.size < period) return prices.average()
    val k = 2.0 / (period + 1.0)
    var ema = prices.take(period).average()
    for (i in period until prices.size) {
      ema = (prices[i] * k) + (ema * (1.0 - k))
    }
    return ema
  }

  private fun calculateRsi(prices: List<Double>, period: Int): Double {
    if (prices.size <= period) return 50.0
    var gains = 0.0
    var losses = 0.0
    for (i in 1..period) {
      val d = prices[i] - prices[i - 1]
      if (d >= 0) gains += d else losses += kotlin.math.abs(d)
    }
    var avgGain = gains / period
    var avgLoss = losses / period
    for (i in (period + 1) until prices.size) {
      val d = prices[i] - prices[i - 1]
      avgGain = (avgGain * (period - 1) + if (d >= 0) d else 0.0) / period
      avgLoss = (avgLoss * (period - 1) + if (d < 0) kotlin.math.abs(d) else 0.0) / period
    }
    if (avgLoss == 0.0) return 100.0
    val rs = avgGain / avgLoss
    return 100.0 - (100.0 / (1.0 + rs))
  }
}

File: lib/services/trading_strategy_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:math' as math;
import 'package:http/http.dart' as http;
import '../models/trade_history.dart';

/// Модуль технического анализа и генерации торговых сигналов Binance Spot Testnet
class TradingStrategyService {
  static const String testnetBaseUrl = 'https://testnet.binance.vision';

  final http.Client _client;
  Timer? _analysisTimer;
  bool _isRunning = false;

  StrategyRiskConfig _config;
  StrategyRiskConfig get config => _config;

  final _signalStreamController = StreamController<SignalData>.broadcast();
  Stream<SignalData> get signalStream => _signalStreamController.stream;

  final _errorStreamController = StreamController<String>.broadcast();
  Stream<String> get errorStream => _errorStreamController.stream;

  Function(SignalData)? onSignal;
  Function(String)? onError;

  TradingStrategyService({
    http.Client? client,
    StrategyRiskConfig? initialConfig,
  })  : _client = client ?? http.Client(),
        _config = initialConfig ?? const StrategyRiskConfig();

  bool get isRunning => _isRunning;

  void updateConfig(StrategyRiskConfig newConfig) {
    _config = newConfig;
  }

  /// Запуск периодического анализа для выбранной пары (по умолчанию каждые 5 сек)
  void start({
    required String symbol,
    Duration interval = const Duration(seconds: 5),
    Function(SignalData)? callback,
  }) {
    stop();
    _isRunning = true;
    onSignal = callback;

    // Первичный запуск без ожидания
    _analyzeSymbol(symbol);

    _analysisTimer = Timer.periodic(interval, (_) {
      if (_isRunning) {
        _analyzeSymbol(symbol);
      }
    });
  }

  void stop() {
    _isRunning = false;
    _analysisTimer?.cancel();
    _analysisTimer = null;
  }

  /// Запрос klines свечей с биржи Binance: GET /api/v3/klines
  Future<List<KlineCandle>> fetchKlines({
    required String symbol,
    String? interval,
    int limit = 100,
  }) async {
    final useInterval = interval ?? _config.interval;
    final uri = Uri.parse(
      '$testnetBaseUrl/api/v3/klines?symbol=${symbol.toUpperCase().trim()}&interval=$useInterval&limit=$limit',
    );

    final res = await _client.get(uri).timeout(const Duration(seconds: 8));
    if (res.statusCode == 200) {
      final List<dynamic> jsonList = jsonDecode(res.body);
      return jsonList.map((e) => KlineCandle.fromJson(e as List<dynamic>)).toList();
    } else {
      String errorMsg = 'HTTP ${res.statusCode}';
      try {
        final err = jsonDecode(res.body);
        if (err is Map && err.containsKey('msg')) {
          errorMsg = '[${err['code']}] ${err['msg']}';
        }
      } catch (_) {}
      throw Exception('Ошибка загрузки klines: $errorMsg');
    }
  }

  /// Запрос klines свечей с биржи и мгновенный расчёт индикаторов/сигнала
  Future<SignalData> fetchAndAnalyze({
    required String symbol,
    String? interval,
    int limit = 60,
  }) async {
    final candles = await fetchKlines(symbol: symbol, interval: interval, limit: limit);
    if (candles.length < 20) {
      throw Exception('Недостаточно свечей klines от Binance для анализа (${candles.length}/20)');
    }

    final indicators = calculateIndicators(candles);
    final currentPrice = candles.last.close;

    final signal = evaluateSignal(
      symbol: symbol,
      currentPrice: currentPrice,
      indicators: indicators,
      candles: candles,
    );

    _signalStreamController.add(signal);
    onSignal?.call(signal);
    return signal;
  }

  /// Расчет индикаторов и проверка условий входа/выхода
  Future<void> _analyzeSymbol(String symbol) async {
    try {
      await fetchAndAnalyze(symbol: symbol);
    } catch (e) {
      final msg = 'Ошибка анализа klines ($symbol): $e';
      _errorStreamController.add(msg);
      onError?.call(msg);
    }
  }

  /// Расчет технических индикаторов: RSI(14), EMA(9), EMA(21), MACD(12,26,9), Volume Avg(20), BB(20,2)
  IndicatorValues calculateIndicators(List<KlineCandle> candles) {
    final closes = candles.map((c) => c.close).toList();
    final volumes = candles.map((c) => c.volume).toList();

    // 1. EMA(9) и EMA(21)
    final ema9List = _calculateEmaSeries(closes, 9);
    final ema21List = _calculateEmaSeries(closes, 21);
    final currentEma9 = ema9List.isNotEmpty ? ema9List.last : closes.last;
    final currentEma21 = ema21List.isNotEmpty ? ema21List.last : closes.last;

    // 2. RSI(14)
    final rsi = _calculateRsi(closes, 14);

    // 3. MACD(12, 26, 9)
    final ema12 = _calculateEmaSeries(closes, 12);
    final ema26 = _calculateEmaSeries(closes, 26);
    final macdSeries = <double>[];
    final minLen = math.min(ema12.length, ema26.length);
    for (int i = 0; i < minLen; i++) {
      macdSeries.add(ema12[i] - ema26[i]);
    }
    final signalSeries = _calculateEmaSeries(macdSeries, 9);
    final currentMacd = macdSeries.isNotEmpty ? macdSeries.last : 0.0;
    final currentSignal = signalSeries.isNotEmpty ? signalSeries.last : 0.0;
    final currentHist = currentMacd - currentSignal;
    final prevHist = (macdSeries.length > 1 && signalSeries.length > 1)
        ? (macdSeries[macdSeries.length - 2] - signalSeries[signalSeries.length - 2])
        : currentHist;

    // 4. Volume: средний за 20 периодов
    final currentVol = volumes.isNotEmpty ? volumes.last : 0.0;
    double avgVol = 0.0;
    final volWindow = volumes.length >= 20 ? volumes.sublist(volumes.length - 20) : volumes;
    if (volWindow.isNotEmpty) {
      avgVol = volWindow.reduce((a, b) => a + b) / volWindow.length;
    }

    // 5. Bollinger Bands (20, 2)
    final bbWindow = closes.length >= 20 ? closes.sublist(closes.length - 20) : closes;
    double bbMiddle = 0.0;
    double bbStd = 0.0;
    if (bbWindow.isNotEmpty) {
      bbMiddle = bbWindow.reduce((a, b) => a + b) / bbWindow.length;
      final sumSquares = bbWindow.map((x) => math.pow(x - bbMiddle, 2)).reduce((a, b) => a + b);
      bbStd = math.sqrt(sumSquares / bbWindow.length);
    }
    final bbUpper = bbMiddle + (bbStd * 2.0);
    final bbLower = bbMiddle - (bbStd * 2.0);

    return IndicatorValues(
      rsi: rsi,
      emaFast: currentEma9,
      emaSlow: currentEma21,
      macdLine: currentMacd,
      macdSignal: currentSignal,
      macdHist: currentHist,
      prevMacdHist: prevHist,
      currentVolume: currentVol,
      avgVolume: avgVol,
      bbUpper: bbUpper,
      bbMiddle: bbMiddle,
      bbLower: bbLower,
    );
  }

  /// Оценка условий и расчет "веса" (score 0-100)
  SignalData evaluateSignal({
    required String symbol,
    required double currentPrice,
    required IndicatorValues indicators,
    required List<KlineCandle> candles,
  }) {
    int longScore = 0;
    int shortScore = 0;
    final List<String> matchedLong = [];
    final List<String> matchedShort = [];

    // Правило 1: RSI (вес 25)
    if (indicators.rsi <= _config.rsiOversold) {
      longScore += 25;
      matchedLong.add('RSI перепродан (${indicators.rsi.toStringAsFixed(1)} <= ${_config.rsiOversold})');
    } else if (indicators.rsi >= _config.rsiOverbought) {
      shortScore += 25;
      matchedShort.add('RSI перекуплен (${indicators.rsi.toStringAsFixed(1)} >= ${_config.rsiOverbought})');
    }

    // Правило 2: EMA тренд (вес 25)
    if (indicators.isEmaBullish) {
      longScore += 25;
      matchedLong.add('EMA Бычий кросс (EMA9: ${indicators.emaFast.toStringAsFixed(1)} > EMA21: ${indicators.emaSlow.toStringAsFixed(1)})');
    } else {
      shortScore += 25;
      matchedShort.add('EMA Медвежий тренд (EMA9 < EMA21)');
    }

    // Правило 3: MACD гистограмма (вес 20)
    if (indicators.isMacdRising) {
      longScore += 20;
      matchedLong.add('MACD гистограмма растёт (${indicators.macdHist.toStringAsFixed(2)})');
    } else {
      shortScore += 20;
      matchedShort.add('MACD гистограмма снижается');
    }

    // Правило 4: Объем выше среднего (вес 15)
    if (indicators.isVolumeAboveAvg) {
      longScore += 15;
      matchedLong.add('Объем свечи выше среднего на 20 периодов');
    }

    // Правило 5: Отскок от нижней границы Боллинджера (вес 15)
    if (indicators.isNearLowerBand) {
      longScore += 15;
      matchedLong.add('Цена коснулась нижней полосы Боллинджера (${indicators.bbLower.toStringAsFixed(1)})');
    } else if (indicators.isNearUpperBand) {
      shortScore += 15;
      matchedShort.add('Цена у верхней полосы Боллинджера (${indicators.bbUpper.toStringAsFixed(1)})');
    }

    // Расчет уровней TP и SL
    final slPrice = currentPrice * (1.0 - (_config.stopLossPercent / 100.0));
    final tpPrice = currentPrice * (1.0 + (_config.takeProfitPercent / 100.0));

    SignalAction action = SignalAction.hold;
    int finalScore = 0;
    List<String> conditions = [];

    if (longScore >= _config.minScoreThreshold) {
      action = SignalAction.buyLong;
      finalScore = longScore;
      conditions = matchedLong;
    } else if (shortScore >= _config.minScoreThreshold) {
      action = SignalAction.sellSpot;
      finalScore = shortScore;
      conditions = matchedShort;
    } else {
      action = SignalAction.hold;
      finalScore = math.max(longScore, shortScore);
      conditions = longScore >= shortScore ? matchedLong : matchedShort;
    }

    return SignalData(
      symbol: symbol,
      action: action,
      score: finalScore,
      currentPrice: currentPrice,
      recommendedStopLoss: slPrice,
      recommendedTakeProfit: tpPrice,
      matchedConditions: conditions,
      indicators: indicators,
      timestamp: DateTime.now(),
    );
  }

  List<double> _calculateEmaSeries(List<double> prices, int period) {
    if (prices.length < period) return [];
    final List<double> result = [];
    final double k = 2.0 / (period + 1.0);

    double sum = 0.0;
    for (int i = 0; i < period; i++) {
      sum += prices[i];
    }
    double ema = sum / period;

    for (int i = 0; i < period - 1; i++) {
      result.add(prices[i]);
    }
    result.add(ema);

    for (int i = period; i < prices.length; i++) {
      ema = (prices[i] * k) + (ema * (1.0 - k));
      result.add(ema);
    }

    return result;
  }

  double _calculateRsi(List<double> prices, int period) {
    if (prices.length <= period) return 50.0;

    double gainSum = 0.0;
    double lossSum = 0.0;

    for (int i = 1; i <= period; i++) {
      final diff = prices[i] - prices[i - 1];
      if (diff >= 0) {
        gainSum += diff;
      } else {
        lossSum += diff.abs();
      }
    }

    double avgGain = gainSum / period;
    double avgLoss = lossSum / period;

    for (int i = period + 1; i < prices.length; i++) {
      final diff = prices[i] - prices[i - 1];
      if (diff >= 0) {
        avgGain = (avgGain * (period - 1) + diff) / period;
        avgLoss = (avgLoss * (period - 1)) / period;
      } else {
        avgGain = (avgGain * (period - 1)) / period;
        avgLoss = (avgLoss * (period - 1) + diff.abs()) / period;
      }
    }

    if (avgLoss == 0) return 100.0;
    final rs = avgGain / avgLoss;
    return 100.0 - (100.0 / (1.0 + rs));
  }

  void dispose() {
    stop();
    _signalStreamController.close();
    _errorStreamController.close();
  }
}
